### ISSUE TITLE
Exception when revoking task, leading to frozen celery worker

### ISSUE THREAD

**Issue by salty-horse** [celery/django-celery-results#88]
This only happens with the `django-db` results backend.

### Versions:
```
celery===4.2.1
django-celery-results==1.0.4
```
### Settings:
```
CELERY_RESULT_BACKEND = 'django-db'
CELERY_BROKER_URL = 'redis://localhost/2'
CELERY_IGNORE_RESULT = False
CELERY_DISABLE_RATE_LIMITS = True
CELERY_WORKER_DIRECT = True
```

### Steps to reproduce:
1. Create a simple task:
```python
from celery import shared_task
@shared_task
def delayed_print(message):
    print(message)
```

2. Run a worker:
`celery -A APPNAME worker -l debug`
```
[config]
.> transport:   redis://localhost:6379/2
.> results:     
.> concurrency: 4 (prefork)
.> task events: OFF (enable -E to monitor tasks in this worker)
```
(BTW, note how the "results:" startup message is empty even if there is a backend).

3. Run a task with a delay, then revoke it:
```
from APPNAME.tasks import delayed_print
result = delayed_print.apply_async(args=['this is a test'], countdown=5)
AsyncResult(result.id).revoke()
```

After 5 seconds, this will cause an exception in the worker:
```
[INFO/MainProcess] Discarding revoked task: APPNAME.tasks.delayed_print[332acfa6-956c-4e00-89d7-b85888940488]
[ERROR/MainProcess] Error in timer: TypeError("delayed_print() missing 1 required positional argument: 'message'",)
Traceback (most recent call last):
  File "/lib/python3.6/site-packages/django/db/models/query.py", line 486, in get_or_create
    return self.get(**lookup), False
  File "/lib/python3.6/site-packages/django/db/models/query.py", line 399, in get
    self.model._meta.object_name
django_celery_results.models.TaskResult.DoesNotExist: TaskResult matching query does not exist.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/lib/python3.6/site-packages/kombu/asynchronous/hub.py", line 145, in fire_timers
    entry()
  File "/lib/python3.6/site-packages/kombu/asynchronous/timer.py", line 68, in __call__
    return self.fun(*self.args, **self.kwargs)
  File "/lib/python3.6/site-packages/celery/worker/consumer/consumer.py", line 477, in apply_eta_task
    self.on_task_request(task)
  File "/lib/python3.6/site-packages/celery/worker/worker.py", line 223, in _process_task_sem
    return self._quick_acquire(self._process_task, req)
  File "/lib/python3.6/site-packages/kombu/asynchronous/semaphore.py", line 62, in acquire
    callback(*partial_args, **partial_kwargs)
  File "/lib/python3.6/site-packages/celery/worker/worker.py", line 228, in _process_task
    req.execute_using_pool(self.pool)
  File "/lib/python3.6/site-packages/celery/worker/request.py", line 518, in execute_using_pool
    if (self.expires or task_id in revoked_tasks) and self.revoked():
  File "/lib/python3.6/site-packages/celery/worker/request.py", line 280, in revoked
    'expired' if expired else 'revoked', False, None, expired,
  File "/lib/python3.6/site-packages/celery/worker/request.py", line 263, in _announce_revoked
    self.id, reason, request=self, store_result=self.store_errors,
  File "/lib/python3.6/site-packages/celery/backends/base.py", line 191, in mark_as_revoked
    traceback=None, request=request)
  File "/lib/python3.6/site-packages/celery/backends/base.py", line 322, in store_result
    request=request, **kwargs)
  File "/lib/python3.6/site-packages/django_celery_results/backends/database.py", line 35, in _store_result
    task_kwargs=task_kwargs,
  File "/lib/python3.6/site-packages/django_celery_results/managers.py", line 50, in _inner
    return fun(*args, **kwargs)
  File "/lib/python3.6/site-packages/django_celery_results/managers.py", line 126, in store_result
    obj, created = self.get_or_create(task_id=task_id, defaults=fields)
  File "/lib/python3.6/site-packages/django/db/models/manager.py", line 82, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/lib/python3.6/site-packages/django/db/models/query.py", line 488, in get_or_create
    return self._create_object_from_params(lookup, params)
  File "/lib/python3.6/site-packages/django/db/models/query.py", line 521, in _create_object_from_params
    params = {k: v() if callable(v) else v for k, v in params.items()}
  File "/lib/python3.6/site-packages/django/db/models/query.py", line 521, in <dictcomp>
    params = {k: v() if callable(v) else v for k, v in params.items()}
  File "/lib/python3.6/site-packages/celery/app/trace.py", line 642, in __protected_call__
    return orig(self, *args, **kwargs)
  File "/lib/python3.6/site-packages/celery/app/task.py", line 375, in __call__
    return self.run(*args, **kwargs)
TypeError: delayed_print() missing 1 required positional argument: 'message'
[INFO/MainProcess] Discarding revoked task: APPNAME.tasks.delayed_print[5fd60b5d-835d-4d25-890d-f2457bb43fc0]
[ERROR/MainProcess] Error in timer: TypeError("delayed_print() missing 1 required positional argument: 'message'",)
Traceback (most recent call last):
  File "/lib/python3.6/site-packages/django/db/models/query.py", line 486, in get_or_create
    return self.get(**lookup), False
  File "/lib/python3.6/site-packages/django/db/models/query.py", line 399, in get
    self.model._meta.object_name
django_celery_results.models.TaskResult.DoesNotExist: TaskResult matching query does not exist.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/lib/python3.6/site-packages/kombu/asynchronous/hub.py", line 145, in fire_timers
    entry()
  File "/lib/python3.6/site-packages/kombu/asynchronous/timer.py", line 68, in __call__
    return self.fun(*self.args, **self.kwargs)
  File "/lib/python3.6/site-packages/celery/worker/consumer/consumer.py", line 477, in apply_eta_task
    self.on_task_request(task)
  File "/lib/python3.6/site-packages/celery/worker/worker.py", line 223, in _process_task_sem
    return self._quick_acquire(self._process_task, req)
  File "/lib/python3.6/site-packages/kombu/asynchronous/semaphore.py", line 62, in acquire
    callback(*partial_args, **partial_kwargs)
  File "/lib/python3.6/site-packages/celery/worker/worker.py", line 228, in _process_task
    req.execute_using_pool(self.pool)
  File "/lib/python3.6/site-packages/celery/worker/request.py", line 518, in execute_using_pool
    if (self.expires or task_id in revoked_tasks) and self.revoked():
  File "/lib/python3.6/site-packages/celery/worker/request.py", line 280, in revoked
    'expired' if expired else 'revoked', False, None, expired,
  File "/lib/python3.6/site-packages/celery/worker/request.py", line 263, in _announce_revoked
    self.id, reason, request=self, store_result=self.store_errors,
  File "/lib/python3.6/site-packages/celery/backends/base.py", line 191, in mark_as_revoked
    traceback=None, request=request)
  File "/lib/python3.6/site-packages/celery/backends/base.py", line 322, in store_result
    request=request, **kwargs)
  File "/lib/python3.6/site-packages/django_celery_results/backends/database.py", line 35, in _store_result
    task_kwargs=task_kwargs,
  File "/lib/python3.6/site-packages/django_celery_results/managers.py", line 50, in _inner
    return fun(*args, **kwargs)
  File "/lib/python3.6/site-packages/django_celery_results/managers.py", line 126, in store_result
    obj, created = self.get_or_create(task_id=task_id, defaults=fields)
  File "/lib/python3.6/site-packages/django/db/models/manager.py", line 82, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/lib/python3.6/site-packages/django/db/models/query.py", line 488, in get_or_create
    return self._create_object_from_params(lookup, params)
  File "/lib/python3.6/site-packages/django/db/models/query.py", line 521, in _create_object_from_params
    params = {k: v() if callable(v) else v for k, v in params.items()}
  File "/lib/python3.6/site-packages/django/db/models/query.py", line 521, in <dictcomp>
    params = {k: v() if callable(v) else v for k, v in params.items()}
  File "/lib/python3.6/site-packages/celery/app/trace.py", line 642, in __protected_call__
    return orig(self, *args, **kwargs)
  File "/lib/python3.6/site-packages/celery/app/task.py", line 375, in __call__
    return self.run(*args, **kwargs)
TypeError: delayed_print() missing 1 required positional argument: 'message'
[INFO/MainProcess] Discarding revoked task: APPNAME.tasks.delayed_print[df83ae59-b455-4928-a4e3-bdbff847aa7f]
[ERROR/MainProcess] Error in timer: TypeError("delayed_print() missing 1 required positional argument: 'message'",)
Traceback (most recent call last):
  File "/lib/python3.6/site-packages/django/db/models/query.py", line 486, in get_or_create
    return self.get(**lookup), False
  File "/lib/python3.6/site-packages/django/db/models/query.py", line 399, in get
    self.model._meta.object_name
django_celery_results.models.TaskResult.DoesNotExist: TaskResult matching query does not exist.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/lib/python3.6/site-packages/kombu/asynchronous/hub.py", line 145, in fire_timers
    entry()
  File "/lib/python3.6/site-packages/kombu/asynchronous/timer.py", line 68, in __call__
    return self.fun(*self.args, **self.kwargs)
  File "/lib/python3.6/site-packages/celery/worker/consumer/consumer.py", line 477, in apply_eta_task
    self.on_task_request(task)
  File "/lib/python3.6/site-packages/celery/worker/worker.py", line 223, in _process_task_sem
    return self._quick_acquire(self._process_task, req)
  File "/lib/python3.6/site-packages/kombu/asynchronous/semaphore.py", line 62, in acquire
    callback(*partial_args, **partial_kwargs)
  File "/lib/python3.6/site-packages/celery/worker/worker.py", line 228, in _process_task
    req.execute_using_pool(self.pool)
  File "/lib/python3.6/site-packages/celery/worker/request.py", line 518, in execute_using_pool
    if (self.expires or task_id in revoked_tasks) and self.revoked():
  File "/lib/python3.6/site-packages/celery/worker/request.py", line 280, in revoked
    'expired' if expired else 'revoked', False, None, expired,
  File "/lib/python3.6/site-packages/celery/worker/request.py", line 263, in _announce_revoked
    self.id, reason, request=self, store_result=self.store_errors,
  File "/lib/python3.6/site-packages/celery/backends/base.py", line 191, in mark_as_revoked
    traceback=None, request=request)
  File "/lib/python3.6/site-packages/celery/backends/base.py", line 322, in store_result
    request=request, **kwargs)
  File "/lib/python3.6/site-packages/django_celery_results/backends/database.py", line 35, in _store_result
    task_kwargs=task_kwargs,
  File "/lib/python3.6/site-packages/django_celery_results/managers.py", line 50, in _inner
    return fun(*args, **kwargs)
  File "/lib/python3.6/site-packages/django_celery_results/managers.py", line 126, in store_result
    obj, created = self.get_or_create(task_id=task_id, defaults=fields)
  File "/lib/python3.6/site-packages/django/db/models/manager.py", line 82, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/lib/python3.6/site-packages/django/db/models/query.py", line 488, in get_or_create
    return self._create_object_from_params(lookup, params)
  File "/lib/python3.6/site-packages/django/db/models/query.py", line 521, in _create_object_from_params
    params = {k: v() if callable(v) else v for k, v in params.items()}
  File "/lib/python3.6/site-packages/django/db/models/query.py", line 521, in <dictcomp>
    params = {k: v() if callable(v) else v for k, v in params.items()}
  File "/lib/python3.6/site-packages/celery/app/trace.py", line 642, in __protected_call__
    return orig(self, *args, **kwargs)
  File "/lib/python3.6/site-packages/celery/app/task.py", line 375, in __call__
    return self.run(*args, **kwargs)
TypeError: delayed_print() missing 1 required positional argument: 'message'
```
The worker will now be frozen. It will continue to print messages of tasks it receives, but will never execute them.
(This action needs to repeat as many times as there are concurrent workers in order to completely stop all operations).

**Comment by salty-horse** [comment]
I found the issue!

In the database backend's [`_store_result()`](https://github.com/celery/django-celery-results/blob/73205bc70f4c6d905d4ccc3d85dcb9cd18739e53/django_celery_results/backends/database.py#L24), `task_name` is always assumed to be either `None` or `str`.
When the task is run successfully, the request contains a `str`.
But when it's asked to be revoked, it contains a `celery.Task` object.

Then, when Django model tries to create the `TaskResult` object in order to modify its status to `REVOKED`, it scans all fields, and if it sees a callable, it calls it. This causes the task to be run! (Unsuccessfully, since there are no arguments).

I hacked around it by adding
```python
from celery import Task
if isinstance(task_name, Task):
    task_name = task_name.name
```
but I would like to know if data that Celery sends is not the real problem. Is it expected that `task` sent in the request is sometimes be a `str` and sometimes the `Task` object?

**Comment by xirdneh** [comment]
This is probably due to a change in Celery's code.
So, it's a good idea to make django-celery-results compatible with this behavior.
Feel free to submit a PR and to add a test for this case, please. We can later review it.
Thanks!

**Comment by kiyoya** [comment]
I think this is the one already fixed in Celery: https://github.com/celery/celery/pull/5068

A new release is not cut though.

**Comment by salty-horse** [comment]
I can confirm that Celery `master` works.

1) Is (current) `master` stable enough to use, or do you recommend applying this specific fix to 4.2.1?
2) From my perspective, this is a breaking bug, especially due to the worker being left in an invalid state. For other users' sake, either a point release of Celery with this fix, or a release of `django-celery-results`, would be appreciated.

**Comment by kiyoya** [comment]
I don't know about Celery's release policy but it looks like they are close to release a new version. See celery/celery#5331

I'm personally using v4.2.1 with the patch.
